lib/puppet_x/puppetlabs/clr_host.rb

```diff
@@ -0,0 +1,510 @@
+          safe_release(@pClrRuntimeHost)
+        end
+
+        start_clr(version)
+      end
+
+      def safe_release(ptr)
+        begin
+          ptr.Release if ptr && !ptr.null?
+          ptr = nil
+        rescue
+        end
+      end
+
+      def FAILED(hr) hr < 0 end
+
+      def hresult_msg(name, code)
+        "#{name} failed (hresult #{format('%#08x', code)})."
+      end
+
+      def start_clr(version)
+        # TODO: old line of code showing a simple CoCreateInstance call
+        # @pITS = COM::TaskScheduler.new
+
+        # simple functions can use existing helper
+        # Load and start .NET runtime
+        FFI::MemoryPointer.new(:pointer) do |ptr|
+          if FAILED(result = CLRCreateInstance(CLSID_CLRMetaHost, IID_ICLRMetaHost, ptr))
+            raise hresult_msg('CLRCreateInstance', result)
+          end
+          @pMetaHost = COM::CLRMetaHost.new(ptr.read_pointer)
+        end
+
+        # Get the ICLRRuntimeInfo corresponding to a particular CLR version. It
+        # supersedes CorBindToRuntimeEx with STARTUP_LOADER_SAFEMODE.
+        FFI::MemoryPointer.new(:pointer) do |ptr|
+          @pMetaHost.GetRuntime(wide_string(version), IID_ICLRRuntimeInfo, ptr)
+          @pRuntimeInfo = COM::CLRRuntimeInfo.new(ptr.read_pointer)
+        end
+
+        # Check if the specified runtime can be loaded into the process. This
+        # method will take into account other runtimes that may already be
+        # loaded into the process and set pbLoadable to TRUE if this runtime can
+        # be loaded in an in-process side-by-side fashion.
+        FFI::MemoryPointer.new(:win32_bool) do |loadable_ptr|
+          @pRuntimeInfo.IsLoadable(loadable_ptr)
+          loadable = loadable_ptr.read_win32_bool
+          if !loadable
+            # TODO: fix this error raising
+            raise Puppet::Error(".NET runtime #{version} cannot be loaded!")
+          end
+        end
+
+        # Load the CLR into the current process and return a runtime interface
+        # pointer. ICorRuntimeHost and ICLRRuntimeHost are the two CLR hosting
+        # interfaces supported by CLR 4.0. Here we demo the ICLRRuntimeHost
+        # interface that was provided in .NET v2.0 to support CLR 2.0 new
+        # features. ICLRRuntimeHost does not support loading the .NET v1.x
+        # runtimes.
+        FFI::MemoryPointer.new(:pointer) do |ptr|
+          @pRuntimeInfo.GetInterface(CLSID_CLRRuntimeHost, IID_ICLRRuntimeHost, ptr)
+          @pClrRuntimeHost = COM::CLRRuntimeHost.new(ptr.read_pointer)
+        end
+
+        @pClrRuntimeHost.Start()
+
+        @pClrRuntimeHost
+      end
+
+      # TODO: demos used a custom C# class CSSimpleObject.GetStringLength
+      # TODO: C# signature must be:
+      # TODO: static int pwzMethodName (String pwzArgument)
+      # TODO: but if we do that, we have to compile one on the fly somehow!
+      def execute_method(assembly_path, type_name, static_method_name, argument)
```

Comment:
@ferventcoder / @jpogran I think this ends up being a dead end.  The method signature is too restrictive, unless we have another way to communicate with the loaded runtime.

No way to pass JSON back without some sort of IPC.